SYCL/ESIMD/api/functional/ctors/ctor_array.cpp

```diff
@@ -0,0 +1,188 @@
+//==------- ctor_array.cpp  - DPC++ ESIMD on-device test -------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu, level_zero
+// XREQUIRES: gpu
+// TODO gpu and level_zero in REQUIRES due to only this platforms supported yet.
+// The current "REQUIRES" should be replaced with "gpu" only as mentioned in
+// "XREQUIRES".
+// UNSUPPORTED: cuda, hip
+// XRUN: %clangxx -fsycl %s -fsycl-device-code-split=per_kernel -o %t.out
+// XRUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: false
+// XFAIL: *
+// TODO Unexpected runtime error "error: unsupported type for load/store" while
+// try to use simd::copy_from(), then simd::copy_to() with fixed-size array that
+// was defined on device side and the test must be enabled when it is resolved.
+//
+// Test for simd constructor from an array.
+// This test uses different data types, dimensionality and different simd
+// constructor invocation contexts.
+// The test does the following actions:
+//  - construct fixed-size array that filled with reference values
+//  - use std::move() to provide it to simd constructor
+//  - bitwise compare expected and retrieved values
+
+#include "common.hpp"
+
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional;
+
+// Descriptor class for the case of calling constructor in initializer context.
+struct initializer {
+  static std::string get_description() { return "initializer"; }
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(DataT(&&ref_data)[NumElems], DataT *const out) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(ref_data)>,
+        "Provided input data is not nonconst rvalue reference");
+    const auto simd_by_init = simd<DataT, NumElems>(std::move(ref_data));
+    simd_by_init.copy_to(out);
+  }
+};
+
+// Descriptor class for the case of calling constructor in variable declaration
+// context.
+struct var_decl {
+  static std::string get_description() { return "variable declaration"; }
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(DataT(&&ref_data)[NumElems], DataT *const out) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(ref_data)>,
+        "Provided input data is not nonconst rvalue reference");
+    simd<DataT, NumElems> simd_by_var_decl(std::move(ref_data));
```

Comment:
```suggestion
    const simd<DataT, NumElems> simd_by_var_decl(std::move(ref_data));
```

Reply:
I've done it in this commit: 1c8437a80d85a14ddc0ec72719271031557b3dda